disks: Be smarter with pre v3.40 bootblocks

andrewbird opened the pull request:
1/ Set drive number on pre v3.40 bootblocks only where we have
a valid drive already, as it seems having the drive number defined
at offset 0x1fd is not as widespread as it first seemed.

2/ Be smarter in mkfatimage BPB modification of existing bootblocks.
This fixes FAT12/FAT16 image file boot with DR-DOS 3.40

stsp commented on src/base/misc/fatfs.c:
```diff
@@ -421,7 +421,9 @@ int read_boot(fatfs_t *f, unsigned char *b)
     } else {
-      b[0x1fd] = f->drive_num;
+      if (b[0x1fd] >= 0x80 && b[0x1fd] <= 0x83) { // Only those that have it
+        b[0x1fd] = f->drive_num;
```
I don't understand this logic.
"If we already have the valid value here,
then oops - overwrite it". Could you please
rearrange the code somehow so that it doesn't
look like an obvious logical error?

andrewbird replied:
I'll explain. In the case where we are reusing a boot.blk provided in a file by the user and it is a version that doesn't have an easily identifiable BPB version i.e. less than v 3.40, we would like to write the drive number into location 0x01fd which is used by some, but not all early versions of DOS. In order to be safe, we test the old boot block to see if that location has valid hard drive number(0x80.. 0x83) already there in that location. if so, we update it with the current drive number. We can't just write it every time without checking, since some boot blocks have other data in that area. On one occasion a boot block had a value of 0x00 in that position which turned out to be the terminator of a string. For that reason we need to be quite narrow about the range of values we could consider to be a valid drive. Have you a better idea how I could accomplish this, or is adding an explanatory comment okay?

stsp replied:
Well, I don't understand.
So if there was a drive number, we put our
own - fine, but if it is not there? Where would
we write the drive number then? There seem
to be no "else" branch, so that logic can't be
valid.

stsp replied:
Also I wonder why there was a line terminator.
0 could have been drive A, not line terminator.

stsp commented on src/tools/periph/mkfatimage16.c:
```diff
@@ -467,24 +467,31 @@ int main(int argc, char *argv[])
+      }
     }
   } else { // Assume we are writing for Dosemu / FreeDOS so v4 BPB
     bpb->v400.hidden_sectors = HIDDEN_SECTORS;
```
Same. Can't we take hidden_sectors from the right place?

andrewbird replied:
We have no struct disk in mkfatimage16, only calculated values. Actually I was toying with the idea of creating a dummy struct disk and using the geometry fixup routine from disks.c to cull some duplicate code.

andrewbird commented on src/tools/periph/mkfatimage16.c:
```diff
@@ -467,24 +467,31 @@ int main(int argc, char *argv[])
+      // v400 && v340
+      bpb->v340.drive_number = 0x80;
+      bpb->v340.flags = 0;
+      bpb->v340.serial_number = 0x12345678;
```
We are only making an image here, so are writing reasonable defaults. When using it later in dosemu it may be fixed up when we know what values to assign, e.g. for drive number etc.

stsp replied:
Yes, sorry, I haven't noticed this is a
different file already (mkfatimage16).

andrewbird commented:
Yes initially I thought 0x00 would be okay, but that offset is not in the BPB it's the final byte before the signature and as such not covered by any spec I could find. On the example I mentioned, setting it to 0x80 caused the boot code to hang, presumably it ran off the end during a copy and overwrote something.
On 10 Dec 2016 20:39, Stas Sergeev <notifications@github.com> wrote:@stsp commented on this pull request.



In src/base/misc/fatfs.c:
> @@ -421,7 +421,9 @@ int read_boot(fatfs_t *f, unsigned char *b)
     if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {
       bpb->v340.drive_number = f->drive_num;
     } else {
-      b[0x1fd] = f->drive_num;
+      if (b[0x1fd] >= 0x80 && b[0x1fd] <= 0x83) { // Only those that have it
+        b[0x1fd] = f->drive_num;

Also I wonder why there was a line terminator.
0 could have been drive A, not line terminator.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/stsp/dosemu2","title":"stsp/dosemu2","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/stsp/dosemu2"}},"updates":{"snippets":[{"icon":"PERSON","message":"@stsp commented on #273"}],"action":{"name":"View Pull Request","url":"https://github.com/stsp/dosemu2/pull/273"}}}

andrewbird commented:
In that case the drive number is being passed via DL register and boot block is not significant in that regard.
On 10 Dec 2016 20:38, Stas Sergeev <notifications@github.com> wrote:@stsp commented on this pull request.



In src/base/misc/fatfs.c:
> @@ -421,7 +421,9 @@ int read_boot(fatfs_t *f, unsigned char *b)
     if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {
       bpb->v340.drive_number = f->drive_num;
     } else {
-      b[0x1fd] = f->drive_num;
+      if (b[0x1fd] >= 0x80 && b[0x1fd] <= 0x83) { // Only those that have it
+        b[0x1fd] = f->drive_num;

Well, I don't understand.
So if there was a drive number, we put our
own - fine, but if it is not there? Where would
we write the drive number then? There seem
to be no "else" branch, so that logic can't be
valid.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/stsp/dosemu2","title":"stsp/dosemu2","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/stsp/dosemu2"}},"updates":{"snippets":[{"icon":"PERSON","message":"@stsp commented on #273"}],"action":{"name":"View Pull Request","url":"https://github.com/stsp/dosemu2/pull/273"}}}

stsp commented:
OK, I see.
So what DOSes actually care about that value
at all? I would assume they all look into DL, so
why do we care? We need to set the geometry,
that's what we do.
The drive num stuff was introduced by commit
2433d402 but the log doesn't say why do we need
that.

stsp commented:
So can we just revert 2433d40 instead and fare well?
I don't suppose it helps a tiny bit, because
if it would, you'd write that into its log msg. :)

andrewbird commented:
Tomorrow I'll remove the drive update for < v3.40 BPB from both mkfatimage and fatfs and retest.  If it makes no difference that drive num differs from boot block to reality on all the DOSes I have, I'll add to this pull request to remove the unnecessary code. However it just seems wrong, to me at least. :-)
On 10 Dec 2016 22:26, Stas Sergeev <notifications@github.com> wrote:So can we just revert 2433d40 instead and fare well?
I don't suppose it helps a tiny bit, because
if it would, you'd write that into its log msg. :)

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/stsp/dosemu2","title":"stsp/dosemu2","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/stsp/dosemu2"}},"updates":{"snippets":[{"icon":"PERSON","message":"@stsp in #273: So can we just revert 2433d40 instead and fare well?\r\nI don't suppose it helps a tiny bit, because\r\nif it would, you'd write that into its log msg. :)"}],"action":{"name":"View Pull Request","url":"https://github.com/stsp/dosemu2/pull/273#issuecomment-266244361"}}}

stsp commented:
Well if you could do that consistently, then maybe.
But as long as you can do that only sometimes, it
might be better to remove. 0x80..0x83 check is not
ultimately reliable either, I'd suppose. So when it
goes to the dirty stuff like this for no reason, it is
better to remove.

Note that we can have a different solution to this.
Assuming that some program can peek into a boot
sector to find out a drive num or the geometry, we
can make it so that mbr will read one boot sector
(eg prepared from boot.blk), and the subsequent
reads can return our boot sector with proper values
filled. This will actually cost just 1-2 extra lines of code,
but I don't think this is necessary.

Btw, I don't see where we fill in 0x1fd when building
our own boot sector. And if it is not there, we should
almost certainly revert 2433d40.

Btw2, I think struct on_disk_bpb should be moved
to fatfs.h.

stsp commented:
> 0x80..0x83 check is not
> ultimately reliable either, I'd suppose

In fact, I am sure its wrong: how does it
account for floppy B? You'll have to swap
drive num from 0 to 1 in that case, and
suddenly you can't.

andrewbird commented:
We don't fill in Ox1fd when building our own boot sector because it's a v4.00 one.

We don't need to account for floppy boot because we don't allow a native boot.blk file, and we also don't build floppy images with mkfatimage16. So ultimately the only drive values here will be Ox80 .. 0x83

I can move on_disk_bpb to fatfs.h as you wish.

andrewbird commented:
Well, that's unexpected! I was coming around to the idea that the code was unnecessary as well.

- I replaced the code that set the byte at 0x1fd to drive_num to always set it to 0x83, an invalid drive. As you can see certain DOSes do use the byte for drive type at boot. It's just luck that most times we don't need to set it as when I manually made the boot blocks for testing they always happened to be based on C:.

~~~
< is this PR code (b[0x1fd] = f->drive_num)
> is test code (b[0x1fd] = 0x083)

< Test DR-DOS-5.00-900615 FAT12 image file ... ok
> Test DR-DOS-5.00-900615 FAT12 image file ... FAIL
< Test DR-DOS-5.00-900615 FAT12 vfs directory boot.blk ... ok
< Test DR-DOS-5.00-900615 FAT16 image file ... ok
> Test DR-DOS-5.00-900615 FAT12 vfs directory boot.blk ... FAIL
> Test DR-DOS-5.00-900615 FAT16 image file ... FAIL
< Test DR-DOS-5.00-900615 FAT16 vfs directory boot.blk ... ok
< Test DR-DOS-5.00-900615 FAT16B image file ... ok
> Test DR-DOS-5.00-900615 FAT16 vfs directory boot.blk ... FAIL
> Test DR-DOS-5.00-900615 FAT16B image file ... FAIL
< Test DR-DOS-5.00-900615 FAT16B vfs directory boot.blk ... ok
> Test DR-DOS-5.00-900615 FAT16B vfs directory boot.blk ... FAIL
< Test DR-DOS-5.00-900814 FAT12 image file ... ok
> Test DR-DOS-5.00-900814 FAT12 image file ... FAIL
< Test DR-DOS-5.00-900814 FAT12 vfs directory boot.blk ... ok
< Test DR-DOS-5.00-900814 FAT16 image file ... ok
> Test DR-DOS-5.00-900814 FAT12 vfs directory boot.blk ... FAIL
> Test DR-DOS-5.00-900814 FAT16 image file ... FAIL
< Test DR-DOS-5.00-900814 FAT16 vfs directory boot.blk ... ok
< Test DR-DOS-5.00-900814 FAT16B image file ... ok
> Test DR-DOS-5.00-900814 FAT16 vfs directory boot.blk ... FAIL
> Test DR-DOS-5.00-900814 FAT16B image file ... FAIL
< Test DR-DOS-5.00-900814 FAT16B vfs directory boot.blk ... ok
> Test DR-DOS-5.00-900814 FAT16B vfs directory boot.blk ... FAIL
< Test DR-DOS-6.00 FAT12 image file ... ok
> Test DR-DOS-6.00 FAT12 image file ... FAIL
< Test DR-DOS-6.00 FAT12 vfs directory boot.blk ... ok
< Test DR-DOS-6.00 FAT16 image file ... ok
> Test DR-DOS-6.00 FAT12 vfs directory boot.blk ... FAIL
> Test DR-DOS-6.00 FAT16 image file ... FAIL
< Test DR-DOS-6.00 FAT16 vfs directory boot.blk ... ok
< Test DR-DOS-6.00 FAT16B image file ... ok
> Test DR-DOS-6.00 FAT16 vfs directory boot.blk ... FAIL
> Test DR-DOS-6.00 FAT16B image file ... FAIL
< Test DR-DOS-6.00 FAT16B vfs directory boot.blk ... ok
> Test DR-DOS-6.00 FAT16B vfs directory boot.blk ... FAIL
< Test MS-DOS-3.20 FAT12 image file ... ok
> Test MS-DOS-3.20 FAT12 image file ... FAIL
< Test MS-DOS-3.20 FAT12 vfs directory boot.blk ... ok
< Test MS-DOS-3.20 FAT16 image file ... ok
> Test MS-DOS-3.20 FAT12 vfs directory boot.blk ... FAIL
> Test MS-DOS-3.20 FAT16 image file ... FAIL
< Test MS-DOS-3.20 FAT16 vfs directory boot.blk ... ok
> Test MS-DOS-3.20 FAT16 vfs directory boot.blk ... FAIL
< Test MS-DOS-3.21 FAT12 image file ... ok
> Test MS-DOS-3.21 FAT12 image file ... FAIL
< Test MS-DOS-3.21 FAT12 vfs directory boot.blk ... ok
< Test MS-DOS-3.21 FAT16 image file ... ok
> Test MS-DOS-3.21 FAT12 vfs directory boot.blk ... FAIL
> Test MS-DOS-3.21 FAT16 image file ... FAIL
< Test MS-DOS-3.21 FAT16 vfs directory boot.blk ... ok
> Test MS-DOS-3.21 FAT16 vfs directory boot.blk ... FAIL
< Test MS-DOS-3.30-Nec FAT12 image file ... ok
> Test MS-DOS-3.30-Nec FAT12 image file ... FAIL
< Test MS-DOS-3.30-Nec FAT12 vfs directory boot.blk ... ok
< Test MS-DOS-3.30-Nec FAT16 image file ... ok
> Test MS-DOS-3.30-Nec FAT12 vfs directory boot.blk ... FAIL
> Test MS-DOS-3.30-Nec FAT16 image file ... FAIL
< Test MS-DOS-3.30-Nec FAT16 vfs directory boot.blk ... ok
> Test MS-DOS-3.30-Nec FAT16 vfs directory boot.blk ... FAIL
< Test MS-DOS-3.30 FAT12 image file ... ok
> Test MS-DOS-3.30 FAT12 image file ... FAIL
< Test MS-DOS-3.30 FAT12 vfs directory boot.blk ... ok
< Test MS-DOS-3.30 FAT16 image file ... ok
> Test MS-DOS-3.30 FAT12 vfs directory boot.blk ... FAIL
> Test MS-DOS-3.30 FAT16 image file ... FAIL
< Test MS-DOS-3.30 FAT16 vfs directory boot.blk ... ok
> Test MS-DOS-3.30 FAT16 vfs directory boot.blk ... FAIL
< Test PC-DOS-3.20-851230 FAT12 image file ... ok
> Test PC-DOS-3.20-851230 FAT12 image file ... FAIL
< Test PC-DOS-3.20-851230 FAT12 vfs directory boot.blk ... ok
< Test PC-DOS-3.20-851230 FAT16 image file ... ok
> Test PC-DOS-3.20-851230 FAT12 vfs directory boot.blk ... FAIL
> Test PC-DOS-3.20-851230 FAT16 image file ... FAIL
< Test PC-DOS-3.20-851230 FAT16 vfs directory boot.blk ... ok
> Test PC-DOS-3.20-851230 FAT16 vfs directory boot.blk ... FAIL
< Test PC-DOS-3.20-860221 FAT12 image file ... ok
> Test PC-DOS-3.20-860221 FAT12 image file ... FAIL
< Test PC-DOS-3.20-860221 FAT12 vfs directory boot.blk ... ok
< Test PC-DOS-3.20-860221 FAT16 image file ... ok
> Test PC-DOS-3.20-860221 FAT12 vfs directory boot.blk ... FAIL
> Test PC-DOS-3.20-860221 FAT16 image file ... FAIL
< Test PC-DOS-3.20-860221 FAT16 vfs directory boot.blk ... ok
> Test PC-DOS-3.20-860221 FAT16 vfs directory boot.blk ... FAIL
< Test PC-DOS-3.30 FAT12 image file ... ok
> Test PC-DOS-3.30 FAT12 image file ... FAIL
< Test PC-DOS-3.30 FAT12 vfs directory boot.blk ... ok
< Test PC-DOS-3.30 FAT16 image file ... ok
> Test PC-DOS-3.30 FAT12 vfs directory boot.blk ... FAIL
> Test PC-DOS-3.30 FAT16 image file ... FAIL
< Test PC-DOS-3.30 FAT16 vfs directory boot.blk ... ok
> Test PC-DOS-3.30 FAT16 vfs directory boot.blk ... FAIL
~~~

- The second test shows the difference when the byte at 0x1fd is not set at all, but the v3.40+ drive num still is (like a partial revert of  commit 2433d40).  I seem to remember that I once had to format /s D: when making boot blocks which is probably the reason I added the code in the first place at commit 2433d40 to fix DR-DOS 5.00 boot

~~~
< is this PR code (b[0x1fd] = f->drive_num)
> is test code with set removed, like partial revert of commit  2433d40

< Test DR-DOS-5.00-900615 FAT12 image file ... ok
> Test DR-DOS-5.00-900615 FAT12 image file ... FAIL
~~~

Here's the boot block
~~~
$ od -A x -t x1z -v -w16 boot-306-4-17.blk
000000 eb 40 90 49 42 4d 20 20 33 2e 33 00 02 08 01 00  >.@.IBM  3.3.....<
000010 02 00 02 37 51 f8 08 00 11 00 04 00 11 00 00 00  >...7Q...........<
000020 00 00 00 00 00 00 00 00 00 00 70 00 ff ff 49 42  >..........p...IB<
000030 4d 42 49 4f 20 20 43 4f 4d 00 50 00 00 08 00 18  >MBIO  COM.P.....<
000040 0a 01 fc 33 c0 8e c0 fa 8e d0 bc 00 7c fb 8a 96  >...3........|...<
000050 fd 01 cd 13 bd 78 00 8b fc c5 76 00 89 7e 00 8c  >.....x....v..~..<
000060 46 02 b9 0b 00 f3 a4 91 8e d8 8b ec c6 46 04 0f  >F............F..<
000070 8a 46 0d 89 46 24 8a 46 10 f7 66 16 03 46 0e 83  >.F..F$.F..f..F..<
000080 d2 00 8b 4e 0b 81 f9 00 02 74 11 72 5e d1 e9 03  >...N.....t.r^...<
000090 c0 83 d2 00 d1 66 24 d1 66 16 eb e9 83 7e 16 0c  >.....f$.f....~..<
0000a0 77 05 c7 46 2c ff 0f 03 46 1c 13 56 1e 50 52 8e  >w..F,...F..V.PR.<
0000b0 46 3a e8 bc 00 5a 58 50 52 8b 5e 11 53 83 c3 0f  >F:...ZXPR.^.S...<
0000c0 b1 04 d3 eb 88 5e 41 8e 46 3c 06 e8 a3 00 07 89  >.....^A.F<......<
0000d0 46 42 89 56 44 59 2b ff 51 57 8d 76 2e b9 0b 00  >FB.VDY+.QW.v....<
0000e0 f3 a6 5f 59 74 13 83 c7 20 e2 ed be cf 7d e8 cd  >.._Yt... ....}..<
0000f0 00 98 cd 16 ea 00 00 ff ff 26 ff 75 1a be c4 7d  >.........&.u...}<
000100 e8 bb 00 59 5a 58 51 8b 4e 16 2b c1 83 da 00 8e  >...YZXQ.N.+.....<
000110 46 3c 51 c6 46 41 01 e8 57 00 59 e2 f5 5b 8e 46  >F<Q.FA..W.Y..[.F<
000120 2a 8b 46 24 88 46 41 53 4b 4b f7 e3 03 46 42 13  >*.F$.FASKK...FB.<
000130 56 44 e8 3c 00 5b 06 8b c3 d1 e3 c4 7e 3c 72 02  >VD.<.[......~<r.<
000140 8e c7 81 7e 2c ff 0f 75 12 03 d8 d1 eb 26 8b 1f  >...~,..u.....&..<
000150 73 04 b1 04 d3 eb 80 e7 0f eb 03 26 8b 1f 07 3b  >s..........&...;<
000160 5e 2c 72 bd 8a 96 fd 01 8a 6e 15 8b 5e 42 ff 6e  >^,r......n..^B.n<
000170 28 33 db 50 52 e8 15 00 8c c0 05 20 00 8e c0 5a  >(3.PR...... ...Z<
000180 58 05 01 00 83 d2 00 fe 4e 41 75 e5 c3 f7 76 18  >X.......NAu...v.<
000190 42 8a ca 33 d2 f7 76 1a 8a f2 d0 cc d0 cc 80 e4  >B..3..v.........<
0001a0 c0 0a cc 8a e8 8a 96 fd 01 b8 01 02 cd 13 73 dc  >..............s.<
0001b0 fe 4e 40 75 f4 e9 33 ff b4 0e 2b db cd 10 ac 84  >.N@u..3...+.....<
0001c0 c0 75 f5 c3 4c 6f 61 64 69 6e 67 2e 2e 2e 00 0d  >.u..Loading.....<
0001d0 0a 43 61 6e 6e 6f 74 20 6c 6f 61 64 20 66 69 6c  >.Cannot load fil<
0001e0 65 0d 0a 50 72 65 73 73 20 61 20 6b 65 79 20 74  >e..Press a key t<
0001f0 6f 20 72 65 74 72 79 0d 0a 00 00 00 00 81 55 aa  >o retry.......U.<
~~~

The DR-DOS boot stalls with message
~~~
Cannot load file
Press a key to retry
~~~

In the boot.log we can see the attempt to access drive 0x81
~~~
INT13: ax=0201 cx=0001 dx=0180                                                  
DISK 80 read [h:1,s:0,t:0](1)->0000:7c00                                        
DISK: /clients/common/dosemu2.git/test-imagedir/drives/c: Trying to read 1 sectors at T/S/H 0/0/1-8704 at pos 0
fatfs: read: dir /clients/common/dosemu2.git/test-imagedir/drives/c, sec 0, len 1
fatfs: dir /clients/common/dosemu2.git/test-imagedir/drives/c, reading boot sector
DISK read @1/0/0 (1) -> 0x7c00 OK.                                              
INT13: ax=0000 cx=0001 dx=0100                                                  
DISK 00 init                                                                    
INT13: ax=0201 cx=0001 dx=0281                                                  
DISK: null dp                                                                   
DISK 81 read [h:2,s:0,t:0](1)->0050:0000                                        
Sector not found 1!                                                             
DISK 81 read [h:2,s:0,t:0](1)->0x500                                            
DISK 81 undefined.                                                              
INT13: ax=0201 cx=0001 dx=0281                                                  
DISK: null dp                                                                   
DISK 81 read [h:2,s:0,t:0](1)->0050:0000                                        
Sector not found 1!                                                             
DISK 81 read [h:2,s:0,t:0](1)->0x500                                            
DISK 81 undefined.                                                              
INT13: ax=0201 cx=0001 dx=0281                                                  
DISK: null dp                                                                   
DISK 81 read [h:2,s:0,t:0](1)->0050:0000                                        
Sector not found 1!                                                             
DISK 81 read [h:2,s:0,t:0](1)->0x500                                            
DISK 81 undefined.                                                              
~~~ 

I think the lesson to come out of this is that I need to document my commits better and state what issue i'm fixing at the time and why. Not to just rely upon my memory when I ask you to pull changes or your ability to magically understand what I'm doing.
So in conclusion I think we need to set the byte where appropriate (testing the existing boot block if < v3.40 for drive in range 0x80..0x83 is the best I can come up with at the moment) and I need to add some more commentary to explain why it looks so odd. Or, perhaps you have a different solution?

stsp commented:
> We don't fill in 0x1fd when building our own boot sector because it's a v4.00 one.

OK but the code in question is:
`    set_geometry(f, b);
    if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {
      bpb->v340.drive_number = f->drive_num;
    } else {
      b[0x1fd] = f->drive_num;
    }
`
Where is the "if (v4.00) do_nothing()" check?
This part must be consistent with what we do manually.
... uh oh, "bpb->v340.signature == 0x29" is actually the
v4.00 check! Man, this is ugly. Could you please add
something like IS_V340(), IS_V400() macros and use
proper unions, like "bpb->v400.signature == 0x29", not
as above.

> We don't need to account for floppy boot because we don't allow a native boot.blk file

Are you sure? I don't see why is so.

> I can move on_disk_bpb to fatfs.h as you wish.

Would be fine.

> I think the lesson to come out of this is that I need to document my commits
> better and state what issue i'm fixing at the time and why.

Yes, every one have to improve on this (with me being the
first offender). Maybe one day I'll come up with the strict
policy on log messages like linux kernel guys already have.

> So in conclusion I think we need to set the byte where appropriate (testing the
> existing boot block if < v3.40 for drive in range 0x80..0x83 is the best I can come
> up with at the moment)

Could you please dump the boot.blk that has 0 there
and hangs once you write a drive num? I wonder if it
is identifiable by some other means.

stsp commented:
> We don't fill in 0x1fd when building our own boot sector because it's a v4.00 one.

Wait, just how is so?
Don't we build any versions of boot sector for
older doses?
In any case, I think we can use f->sys_type even
when dealing with pre-defined boot.blk. While this
may look wrong, it is in fact correct because the
arbitrary boot.blk can only load whatever fatfs.c
puts to the system area (starting from cluster 2)
and not anything else. If it wants to load something
else than f->sys_type allows, it will not work.

andrewbird commented:
- Here's a DR-DOS 3.41 boot block that has the string terminator in the 0x1fd byte. Strangely it didn't crash this time after altering it to 0x80. Perhaps it depends upon what seemingly random data is following...
~~~
tar -xvf test-binaries/DR-DOS-3.41.tar boot-306-4-17.blk && od -A x -t x1z -v -w16 boot-306-4-17.blk
boot-306-4-17.blk
000000 eb 3c 90 49 42 4d 20 20 33 2e 32 00 02 08 01 00  >.<.IBM  3.2.....<
000010 02 00 02 37 51 f8 08 00 11 00 04 00 11 00 00 00  >...7Q...........<
000020 00 00 00 00 80 00 00 70 00 00 ff 0f 01 00 08 00  >.......p........<
000030 08 00 44 52 42 49 4f 53 20 20 53 59 53 00 fc 8c  >..DRBIOS  SYS...<
000040 c8 8e d0 8e d8 bc 00 7c 8b ec fb 99 8a 46 10 f7  >.......|.....F..<
000050 66 2e 03 46 2c 03 46 1c 13 56 1e 8b 5e 11 b1 04  >f..F,.F..V..^...<
000060 d3 eb 2b ff 89 46 00 89 56 02 01 5e 00 11 7e 02  >..+..F..V..^..~.<
000070 b9 50 00 8e c1 c6 46 04 01 e8 ec 00 b9 00 08 8e  >.P....F.........<
000080 c1 88 5e 04 e8 e1 00 8b 4e 11 51 57 8d 76 32 b9  >..^.....N.QW.v2.<
000090 0b 00 f3 a6 5f 59 74 10 83 c7 20 e2 ed 56 be cb  >...._Yt... ..V..<
0000a0 01 e8 01 01 5e e9 b1 00 26 8b 4d 1a 89 4e 40 8b  >....^...&.M..N@.<
0000b0 4e 2e 2b c1 83 da 00 c6 46 04 40 8c c3 53 51 e8  >N.+.....F.@..SQ.<
0000c0 a6 00 59 81 c3 00 08 8e c3 05 40 00 83 d2 00 83  >..Y.......@.....<
0000d0 e9 40 7f ea 07 8b 46 30 b1 05 d3 e0 89 46 07 49  >.@....F0.....F.I<
0000e0 8b 5e 40 8b 76 27 56 8b c3 2b d2 48 48 f7 66 30  >.^@.v'V..+.HH.f0<
0000f0 03 46 00 13 56 02 50 52 2b ff 03 7e 30 03 76 07  >.F..V.PR+..~0.v.<
000100 06 80 ff 80 72 05 b8 00 18 8e c0 8b c3 d1 e3 f6  >....r...........<
000110 46 29 ff 75 12 03 d8 d1 eb a8 01 26 8b 1f 74 02  >F).u.......&..t.<
000120 d3 eb 80 e7 0f eb 03 26 8b 1f 07 40 3b c3 74 ca  >.......&...@;.t.<
000130 57 3b 5e 2a 72 b0 be f1 01 e8 69 00 3b ec 74 0b  >W;^*r.....i.;.t.<
000140 8f 46 04 5a 58 07 e8 1f 00 eb f1 8a 56 24 2a f6  >.F.ZX.......V$*.<
000150 8a 6e 15 8b 5e 00 ff 6e 25 e8 49 00 be b4 01 e8  >.n..^..n%.I.....<
000160 43 00 2b c0 cd 16 cd 19 50 52 53 f7 76 18 42 8a  >C.+.....PRS.v.B.<
000170 ca 33 d2 f7 76 1a 8a f2 d0 cc d0 cc 80 e4 c0 0a  >.3..v...........<
000180 cc 8a e8 8a 46 04 b4 02 c6 46 06 05 8a 56 24 33  >....F....F...V$3<
000190 db 50 cd 13 58 73 0a fe 4e 06 75 f3 be e0 01 eb  >.P..Xs..N.u.....<
0001a0 b8 5b 5a 58 c3 03 f5 ac 84 c0 74 f8 b4 0e 2b db  >.[ZX......t...+.<
0001b0 cd 10 eb f3 0d 0a 50 72 65 73 73 20 61 20 6b 65  >......Press a ke<
0001c0 79 20 74 6f 20 72 65 74 72 79 00 0d 43 61 6e 6e  >y to retry..Cann<
0001d0 6f 74 20 66 69 6e 64 20 66 69 6c 65 20 2d 20 00  >ot find file - .<
0001e0 0d 44 69 73 6b 20 72 65 61 64 20 65 72 72 6f 72  >.Disk read error<
0001f0 00 0d 4c 6f 61 64 69 6e 67 20 2e 2e 2e 00 55 aa  >..Loading ....U.<
000200
~~~

- I think you are correct about being able to use a native boot.blk file on virtual floppy from directory. I hadn't spotted that was possible. So yes in this case the boot block should be updated with proper drive number, not just if existing value in range of 0x80..0x83. I guess I should be testing that too!

- Currently Dosemu always makes its own v4.0 boot block for virtual directory boots, I don't see why it should make anything else since the boot block code it executes is its own. But please correct me if I'm wrong?

- If we can narrow the pre v3.40 bootblocks that don't have drive number in byte 0x1fd to just DR-DOS 3.40 / 3.41 then setting it on sys_type of OLDDRD_D can be avoided in fatfs.c.

- We can't avoid writing the drive number on the DR-DOS 3.40 / 3.41 boot block as above when making an image in mkfatimage16.c as we don't know the running system type. Since we don't have to cater for floppy values 0x00 and 0x01 (we are only making hard disk images),  perhaps the 0x80..0x83 check will suffice there? But if you expect to be able to use a boot block from a floppy to boot a hard disk then that logic would fail.

- Happy to add IS_BPB_V400 macros etc, but since some setting code is common to both v3.40 and 4.00 do you want that duplicated. e.g from mkfatimage16.c .
```C
    if(bpb->v340.signature == 0x28 || bpb->v340.signature == 0x29) {            
      if(bpb->v340.signature == 0x29) { // v400                                 
        memset(bpb->v400.vol_label, ' ', 11);                                   
        memcpy(bpb->v400.vol_label, volume_label, strlen(volume_label));        
        memcpy(bpb->v400.fat_type,                                              
               p_type == P_TYPE_12BIT ? "FAT12   " : "FAT16   ", 8);            
      }                                                                         
      // v400 && v340                                                           
      bpb->v340.drive_number = 0x80;                                            
      bpb->v340.flags = 0;                                                      
      bpb->v340.serial_number = 0x12345678;                                     
                                                                                
      bpb->v340.hidden_sectors = HIDDEN_SECTORS;                                
      bpb->v340.num_sectors_large = (p_sectors < 65536L) ? 0 : p_sectors;
  } else { // do pre 3.40 stuff
  }
```
becomes
```C
    if (IS_BPB_V400(bpb)) { 
        memset(bpb->v400.vol_label, ' ', 11);                                   
        memcpy(bpb->v400.vol_label, volume_label, strlen(volume_label));        
        memcpy(bpb->v400.fat_type,                                              
               p_type == P_TYPE_12BIT ? "FAT12   " : "FAT16   ", 8);

        bpb->v400.drive_number = 0x80;                                            
        bpb->v400.flags = 0;                                                      
        bpb->v400.serial_number = 0x12345678;
        bpb->v400.hidden_sectors = HIDDEN_SECTORS;                                
        bpb->v400.num_sectors_large = (p_sectors < 65536L) ? 0 : p_sectors;
   } else if (IS_BPB_V340(bpb)) {
        bpb->v340.drive_number = 0x80;                                            
        bpb->v340.flags = 0;                                                      
        bpb->v340.serial_number = 0x12345678;
        bpb->v340.hidden_sectors = HIDDEN_SECTORS;                                
        bpb->v340.num_sectors_large = (p_sectors < 65536L) ? 0 : p_sectors; 
   } else { // do pre 3.40 stuff
   }
```
Otherwise I'd be setting v400 fields via bpb->v340, which I presume you didn't like?

stsp commented:
> I don't see why it should make anything else since the boot block code it executes is its own.

This is very strange to hear given that you claim DOSes
peek into a boot sector for various things, like the drive
number at 1fd. So if we do our own boot sector, DOS doesn't
peek its values and it perfectly fine with seeing the 4.00 layout?

> We can't avoid writing the drive number on the DR-DOS 3.40 / 3.41 boot
> block as above when making an image in mkfatimage16.c as we don't know
> the running system type.

We can still find that out by looking into the boot sector's
system file name (DRBIOS.SYS in this case). Or we can
add a parameter allowing the user to specify DOS version.
But in any case, if you want that heuristic in mkfatimage16
then fine. I wouldn't care much, but lets at least not put it
into fatfs.c.

> Otherwise I'd be setting v400 fields via bpb->v340, which I presume you didn't like?

Would it be possible to make common structs so that
v400 to have the v340 at start, etc? ms extensions in
gcc allow us to add anonymous fields, so if you start
your v400 struct with just `v340;` then all v340's fields
will be accessible without specifying extra member.

stsp commented:
With common members you can do something like
`bpb->v340_400.drive_num = xx;`
and use ->v400 only for specific fields.

andrewbird commented:
> This is very strange to hear given that you claim DOSes
> peek into a boot sector for various things, like the drive
> number at 1fd. So if we do our own boot sector, DOS doesn't
> peek its values and it perfectly fine with seeing the 4.00 layout?

No, In some cases I think the boot sector code looks at the value at 0x1fd, rather than figuring out which drive it is booting from, before starting DOS and informing it which drive via DL etc.
I have an example (issue #275) In which DR-DOS 3.41 boots fat16b just fine from directory with the v4.00 boot block that dosemu makes, but can't boot fat16b from either image or directory when the native v3.20 boot block is used.

> With common members you can do something like
> bpb->v340_400.drive_num = xx;
> and use ->v400 only for specific fields.

Thanks I'll look into that, or if you want you do it and I'll rebase?

stsp commented:
> No, In some cases I think the boot sector code looks at the value at 0x1fd, rather
> than figuring out which drive it is booting from, before starting DOS and informing
> it which drive via DL etc.

OK, so how does the problematic boot sector work?
As I understand, it is pre-3.20, and yet does not use 1fd.
Where does it give the DL value then?

> Thanks I'll look into that, or if you want you do it and I'll rebase?

I'll see, but you can beat me on this. :)

stsp commented:
> Thanks I'll look into that, or if you want you do it and I'll rebase?

Done.

> I can move on_disk_bpb to fatfs.h as you wish.

Probably not as long as mkfatimage16 also uses it...
fatfs.h appeared to be too internal.
Of course we can re-use include/fatfs.h that currently
has just a single line, but I wonder if it worth the trouble.
Whatever you think is better.

andrewbird commented:
I force pushed a revision to this PR
I didn't update against your new BPB structure yet, as although it now looks better, I'm still not sure if the changes to it make it better. I thought about the funky v340_v400 suggestion, and I think it makes the structure less clear than it is now in devel. On the whole, I'm happy with the BPB structure as it is in devel without any changes, except for addition of the macros. See what you think?

stsp commented on src/base/misc/fatfs.c:
```diff
@@ -399,12 +425,30 @@ static void set_geometry(fatfs_t *f, unsigned char *b)
+  } else if (f->sys_type == OLDDRD_D) { // DR-DOS 3.40 / 3.41 has v3.20 BPB
+    bpb->v320.hidden_sectors = f->hidden_secs;
+    bpb->v320.num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
+    // drive_number not passed in boot block as string data in 0x1fd position
```
So where is it passed from?

andrewbird replied:
My feeling is that it's not passed, it's discovered in the boot code somehow. After all it's only the drive number that was correct at the time the file system was made. It would seem disassembling the real boot block, or perhaps doing format /s on D: under DR-DOS 3.41 and comparing is the only way to find out.

stsp commented on src/base/misc/fatfs.c:
```diff
@@ -417,27 +461,23 @@ int read_boot(fatfs_t *f, unsigned char *b)
          f->fat_type == FAT_TYPE_FAT12 ? "FAT12   " : "FAT16   ", 8);
+  }
+
+  set_geometry(f, b);
```
You assign drive number in set_geometry() now,
plus it is also assigned on line 471. I wonder why
moving the drive number to set_geometry()?

stsp commented:
> I thought about the funky v340_v400 suggestion, and I think it makes the structure less clear

How about my latest patch, plus simply prefixing the
relevant field names with the v340_v400 prefix? Like
`v340_400_signature`, `v400_vol_label` etc?
Looks like this will combine the simplicity of my approach
with the "safety feel" of yours.

andrewbird commented:
Your latest BPB struct plus v340_400_signature etc would be reasonable, except for hidden_sectors and num_sectors_large as they are the same as v331. How would you deal with setting hidden_sectors on a v400 BPB?
```C
bpb->v331.hidden_sectors // seems like an obvious bug when in the v400 section (where we started!)
```
or rename v331 internal struct to 
```C
bpb->v331_v340_v400.hidden_sectors
```
Seems like the struct user has to know about field version compatibility and it's not intuitive. Anyway I figure we've probably spent too much time debating field naming and it's probably not worth it :)

stsp commented:
`bpb->v331_400.hidden_sectors`?
Note that I do not put 'v' before '400' implying the
v[range] syntax. The 'v331_v400' would instead
mean only 331 and 400, not the range.

andrewbird commented:
Ahh, okay a subtle distinction. Do you want to fixup your commit and I'll test it before rebasing my patch just so that we know it didn't break anything?

stsp commented:
> Do you want to fixup your commit and I'll test it 

Done.

andrewbird commented:
Cool, test status of fatfs_bpb branch == current devel - good to merge.

stsp commented:
Done. :)

andrewbird commented:
Okay I rebased on updated devel. Tests run same as before i.e. FAT12 & FAT16 image boots fixed on DR-DOS 3.40.

There is one further thought I have regarding distinguishing v3.20 from v3.31 BPBs in existing boot blocks, and that is to read the existing v331_400_hidden_sectors value and compare with v320.hidden_sectors. I have to work out if the little endianness and type difference is going give a feasible test. I'm not sure if this is worth pursuing?

stsp commented:
Thanks!

> v320.hidden_sectors. I have to work out if the little endianness and type difference
> is going give a feasible test. I'm not sure if this is worth pursuing?

You mean to get rid of sys_type check in fatfs.c, or
for better heuristic in mkfatimage16? I think sys_type
is fine in fatfs.c, and as for mkfatimage - I don't expect
anyone to use the pre-built boot blocks there. So I don't
see the compelling reason for such heuristic, but if it works
good (which I doubt) and you think it fixes something, you
can try it out.

andrewbird commented:
I think the sys_type check is fine for DR-DOS, but of course does nothing for any MS-DOS or PC-DOS that uses a v3.20 BPB. I suspect that my current tests don't show up any problems since I'm using a native boot.blk file that has the same type1/2/9 geometry as the fatfs directory under test. Probably a better test would be to use the first sector off the boot-floppy and see if Dosemu can convert its geometry properly. Anyway I figure I should get the current crop of failing tests fixed first before inventing new ones :)

stsp commented:
> better test would be to use the first sector off the boot-floppy and see if Dosemu can
> convert its geometry 

It would be interesting of course, but I think boot.blk
is mostly for testing things. It allows to start from something
functional while adding a support for new doses. But I don't
think it is going to be used by someone else than yourself, so
maybe you can prioritize other interesting things. :)
